[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74561 [Dynamic RPC] Add graceful shutdown for dynamic RPC members
* #74035 [Dynamic RPC] Allow existing ranks to communicate with newly joined ranks
* **#73373 [Dynamic RPC] Allow newly joined ranks to communicate with existing ranks**
* #73372 [Dynamic RPC] Allow for optional world_size argument in init_rpc

This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

Test:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

Differential Revision: [D35052544](https://our.internmc.facebook.com/intern/diff/D35052544)

H-Huang committed:
[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.h:
```diff
@@ -16,6 +16,15 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+// put its own worker info in the store and discover all the ranks that came
+// before it/
+std::unordered_map<std::string, worker_id_t> collectCurrentNames(
+    ::c10d::PrefixStore store,
```
curious, why this is a `PrefixStore` instead of `const c10::intrusive_ptr<::c10d::Store>&`?

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -41,6 +41,75 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
+  size_t end;
+  while ((end = s.find(delim, start)) != std::string::npos) {
+    tokens.push_back(s.substr(start, end - start));
```
does `emplace_back` help here?

H-Huang replied:
Thanks, updated

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -41,6 +41,75 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+    tokens.push_back(s.substr(start, end - start));
+    start = end + delim.length();
+  }
+  tokens.push_back(s.substr(start));
```
ditto

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.h:
```diff
@@ -16,6 +16,15 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+// name-to-id mapping for the current peers in the group. The current rank will
+// put its own worker info in the store and discover all the ranks that came
+// before it/
+std::unordered_map<std::string, worker_id_t> collectCurrentNames(
```
this also needs to be called with the token held, right? If yes, let's highlight that in the comments

H-Huang replied:
Yes, added a comment for that.

mrshenli commented on torch/csrc/distributed/rpc/agent_utils.cpp:
```diff
@@ -41,6 +41,75 @@ std::unordered_map<std::string, worker_id_t> collectNames(
+  std::vector<uint8_t> selfNameVector(
+      (uint8_t*)selfName.c_str(),
+      (uint8_t*)selfName.c_str() + selfName.length());
+  store.set(c10::to_string(selfId), selfNameVector);
```
do we need to verify that id does not already exist?

mrshenli reviewed (commented):
LGTM. Please add some tests to cover new code.

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

TODO: add test for the above scenario

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

Test:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

[ghstack-poisoned]

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

Test:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35052544).

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4341,6 +4341,42 @@ def test_init_rpc_without_world_size(self):
+            result = rpc.rpc_sync(worker_name(0), torch.add, args=(torch.tensor(1), torch.tensor(1)))
+            self.assertEqual(torch.add(torch.tensor(1), torch.tensor(1)), result)
+
+        # TODO: Sync before shutdown since graceful shutdown is not yet implemented
```
By "TODO", I assume you mean we should be able to remove this when dynamic group is fully ready?

H-Huang replied:
Yes! I updated the comment to be more clear

mrshenli reviewed (approved):
LGTM! Thanks Howard!

H-Huang committed:
Update on "[Dynamic RPC] Allow newly joined ranks to communicate with existing ranks"


This PR allows for newly joined ranks in Dynamic RPC to communicate with ranks that have already joined the group. That is, rank N will be able to run RPC against all ranks <= N.

Previously:

Process 1 (init):
```python
init_rpc("worker0", rank=0)
```
Process2 (command against a rank that already joined, would fail):
```python
init_rpc("worker1", rank=1)
rpc.sync("worker0", torch.add, (torch.tensor(1), torch.tensor(1)))
```

Now:
Above scenario succeeds

Test:
`pytest test/distributed/rpc/test_tensorpipe_agent.py -vsk test_init_rpc_without_world_size`

Differential Revision: [D35052544](https://our.internmc.facebook.com/intern/diff/D35052544)

[ghstack-poisoned]

H-Huang commented:
@H-Huang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35052544).
